ops/testing.py

```diff
@@ -834,5 +857,332 @@ def network_get(self, endpoint_name, relation_id=None):
     def add_metrics(self, metrics, labels=None):
         raise NotImplementedError(self.add_metrics)
 
-    def get_pebble(self, container_name):
-        raise NotImplementedError(self.get_pebble)
+    def juju_log(self, level, msg):
+        raise NotImplementedError(self.juju_log)
+
+    def get_pebble(self, socket_path: str):
+        client = self._pebble_clients.get(socket_path, None)
+        if client is None:
+            client = _TestingPebbleClient(self)
+            self._pebble_clients[socket_path] = client
+        return client
+
+
+class _TestingPebbleClient:
+    """This conforms to the interface for pebble.Client but provides canned data.
+
+    DO NOT use this class directly, it is used by `Harness`_ to run interactions with Pebble.
+    `Harness`_ is responsible for maintaining the internal consistency of the values here,
+    as the only public methods of this type are for implementing Client.
+    """
+
+    def __init__(self, backend: _TestingModelBackend):
+        self._backend = _TestingModelBackend
+        self._layers = {}
+        # Has a service been started/stopped?
+        self._service_status = {}
+
+    def get_system_info(self) -> pebble.SystemInfo:
+        """Get system info."""
+        raise NotImplementedError(self.get_system_info)
```

Comment:
Out of curiosity, why `self.get_system_info` and not `'get_system_info'`?

Reply:
get_system_info is an unbound variable. self.get_system_info has the access to the actual function object.

Reply:
Sorry, I was suggesting a string, so `raise NotImplementedError('get_system_info')`. But I'm totally happy with what you have; that's very nitty.